src/http/h1/decode.rs

```diff
@@ -136,113 +126,183 @@ impl Decoder {
     }
 }
 
-fn eat<R: Read>(rdr: &mut R, bytes: &[u8]) -> io::Result<()> {
-    let mut buf = [0];
-    for &b in bytes.iter() {
-        match try!(rdr.read(&mut buf)) {
-            1 if buf[0] == b => (),
+macro_rules! byte (
+    ($rdr:ident) => ({
+        let mut buf = [0];
+        match try!($rdr.read(&mut buf)) {
+            1 => buf[0],
             _ => return Err(io::Error::new(io::ErrorKind::InvalidInput,
-                                          "Invalid characters found")),
+                                                "Invalid chunk size line")),
         }
-    }
-    Ok(())
-}
+    })
+);
 
-/// Chunked chunks start with 1*HEXDIGIT, indicating the size of the chunk.
-fn read_chunk_size<R: Read>(rdr: &mut R) -> io::Result<u64> {
-    macro_rules! byte (
-        ($rdr:ident) => ({
-            let mut buf = [0];
-            match try!($rdr.read(&mut buf)) {
-                1 => buf[0],
-                _ => return Err(io::Error::new(io::ErrorKind::InvalidInput,
-                                                  "Invalid chunk size line")),
-
-            }
+impl ChunkedState {
+    fn step<R: Read>(
+            &self,
+            body: &mut R,
+            size: &mut u64,
+            buf: &mut [u8],
+            read: &mut usize) -> io::Result<ChunkedState> {
+        Ok(match *self {
+            ChunkedState::Size => try!(ChunkedState::read_size(body, size)),
+            ChunkedState::SizeLws => try!(ChunkedState::read_size_lws(body)),
+            ChunkedState::SizeLf => try!(ChunkedState::read_size_lf(body, size)),
+            ChunkedState::Body => try!(ChunkedState::read_body(body, size, buf, read)),
+            ChunkedState::BodyCr => try!(ChunkedState::read_body_cr(body)),
+            ChunkedState::BodyLf => try!(ChunkedState::read_body_lf(body)),
+            ChunkedState::End => panic!("Calling step on End state"),
```

Comment:
I have observed this `panic`. I need to investigate how `Decoder` is called and formulate a test for it.